File: long_questions/DSA/Leetcode/golang/13_Union_Find/128_Longest_Consecutive_Sequence.go
package main

import "fmt"

// 128. Longest Consecutive Sequence
// Time: O(N), Space: O(N)
func longestConsecutive(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	
	numSet := make(map[int]bool)
	for _, num := range nums {
		numSet[num] = true
	}
	
	maxLength := 0
	
	for num := range numSet {
		// Only start counting from the beginning of a sequence
		if !numSet[num-1] {
			currentNum := num
			currentLength := 1
			
			// Count the length of the consecutive sequence
			for numSet[currentNum+1] {
				currentNum++
				currentLength++
			}
			
			if currentLength > maxLength {
				maxLength = currentLength
			}
		}
	}
	
	return maxLength
}

// Union-Find approach
func longestConsecutiveUnionFind(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	
	// Initialize Union-Find
	parent := make(map[int]int)
	rank := make(map[int]int)
	
	for _, num := range nums {
		parent[num] = num
		rank[num] = 0
	}
	
	var find func(int) int
	find = func(x int) int {
		if parent[x] != x {
			parent[x] = find(parent[x])
		}
		return parent[x]
	}
	
	var union func(int, int)
	union = func(x, y int) {
		rootX := find(x)
		rootY := find(y)
		
		if rootX == rootY {
			return
		}
		
		if rank[rootX] < rank[rootY] {
			parent[rootX] = rootY
		} else if rank[rootX] > rank[rootY] {
			parent[rootY] = rootX
		} else {
			parent[rootY] = rootX
			rank[rootX]++
		}
	}
	
	// Union consecutive numbers
	for _, num := range nums {
		if _, exists := parent[num-1]; exists {
			union(num, num-1)
		}
		if _, exists := parent[num+1]; exists {
			union(num, num+1)
		}
	}
	
	// Count the size of each component
	componentSize := make(map[int]int)
	for _, num := range nums {
		root := find(num)
		componentSize[root]++
	}
	
	maxLength := 0
	for _, size := range componentSize {
		if size > maxLength {
			maxLength = size
		}
	}
	
	return maxLength
}

// Sorting approach (O(N log N))
func longestConsecutiveSorting(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	
	// Remove duplicates and sort
	unique := make(map[int]bool)
	for _, num := range nums {
		unique[num] = true
	}
	
	sorted := make([]int, 0, len(unique))
	for num := range unique {
		sorted = append(sorted, num)
	}
	
	// Simple bubble sort for demonstration (in practice, use sort.Ints)
	for i := 0; i < len(sorted)-1; i++ {
		for j := 0; j < len(sorted)-i-1; j++ {
			if sorted[j] > sorted[j+1] {
				sorted[j], sorted[j+1] = sorted[j+1], sorted[j]
			}
		}
	}
	
	maxLength := 1
	currentLength := 1
	
	for i := 1; i < len(sorted); i++ {
		if sorted[i] == sorted[i-1]+1 {
			currentLength++
		} else {
			if currentLength > maxLength {
				maxLength = currentLength
			}
			currentLength = 1
		}
	}
	
	if currentLength > maxLength {
		maxLength = currentLength
	}
	
	return maxLength
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: HashSet-based Sequence Detection
- **HashSet Storage**: Store all numbers for O(1) lookup
- **Sequence Start Detection**: Only start counting from sequence beginnings
- **Linear Expansion**: Expand forward from each sequence start
- **Maximum Tracking**: Track longest sequence found

## 2. PROBLEM CHARACTERISTICS
- **Unordered Array**: Input array with no specific order
- **Consecutive Numbers**: Find longest sequence of consecutive integers
- **Unique Elements**: Duplicates don't affect sequence length
- **Linear Time**: Need O(N) solution (sorting would be O(N log N))

## 3. SIMILAR PROBLEMS
- Longest Consecutive Sequence in Binary Tree (LeetCode 298)
- Consecutive Numbers Sum (LeetCode - variant)
- Find Missing Numbers (LeetCode - variant)
- Array Partition into Consecutive Subsets (LeetCode - variant)

## 4. KEY OBSERVATIONS
- **HashSet advantage**: O(1) lookup for number existence
- **Sequence start condition**: num-1 not in set indicates sequence start
- **Forward expansion**: Only need to check num+1, num+2, etc.
- **Duplicate handling**: HashSet automatically handles duplicates

## 5. VARIATIONS & EXTENSIONS
- **Longest Consecutive Subarray**: Consecutive in original order
- **K-length Consecutive Sequences**: Find all sequences of length K
- **Circular Consecutive**: Treat array as circular
- **Multiple Arrays**: Find longest across multiple arrays

## 6. INTERVIEW INSIGHTS
- Always clarify: "Can the array be empty? Are there duplicates?"
- Edge cases: empty array, single element, all duplicates
- Time complexity: O(N) - each number processed once
- Space complexity: O(N) - for HashSet storage

## 7. COMMON MISTAKES
- Not using HashSet and trying O(N²) approach
- Starting sequence from every number instead of just starts
- Not handling empty array case
- Forgetting to handle duplicates properly
- Using sorting (O(N log N)) when O(N) is possible

## 8. OPTIMIZATION STRATEGIES
- **Current solution is optimal**: O(N) time, O(N) space
- **Union Find alternative**: More complex but demonstrates pattern
- **Bit manipulation**: Not applicable for this problem
- **Early termination**: Not applicable (need to check all numbers)

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like finding the longest chain of consecutive numbers:**
- You have a bag of numbered balls (unordered array)
- You want to find the longest chain where each number is one more than the previous
- Instead of checking every possible starting point, only start chains at the beginning
- A number is the beginning if the previous number doesn't exist
- Once you find a beginning, count how long the chain goes forward

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Unordered array of integers
2. **Goal**: Find length of longest consecutive sequence
3. **Output**: Integer representing maximum sequence length
4. **Constraint**: Must be O(N) time complexity

#### Phase 2: Key Insight Recognition
- **"HashSet necessity"** → Need O(1) lookup to achieve O(N) time
- **"Sequence start detection"** → Only start from numbers where num-1 doesn't exist
- **"Forward expansion"** → Once at sequence start, count forward consecutively
- **"Duplicate handling"** → HashSet automatically handles duplicates

#### Phase 3: Strategy Development
```
Human thought process:
"I need to find the longest consecutive sequence.
If I sort, it's O(N log N), but I need O(N).
I'll use a HashSet for O(1) lookups.
For each number, I'll check if it's the start of a sequence.
If num-1 doesn't exist, it's a sequence start.
From there, I'll count how far the sequence goes forward."
```

#### Phase 4: Edge Case Handling
- **Empty array**: Return 0
- **Single element**: Return 1
- **All duplicates**: Return 1
- **All consecutive**: Return array length (deduplicated)

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Array: [100, 4, 200, 1, 3, 2]

Human thinking:
"I'll build a HashSet: {100, 4, 200, 1, 3, 2}

Now check each number:
100: Is 99 in set? No, so 100 is sequence start
- Count: 100 (101? No) → length = 1

4: Is 3 in set? Yes, so 4 is not sequence start

200: Is 199 in set? No, so 200 is sequence start  
- Count: 200 (201? No) → length = 1

1: Is 0 in set? No, so 1 is sequence start
- Count: 1, 2, 3, 4 (5? No) → length = 4

3: Is 2 in set? Yes, so 3 is not sequence start
2: Is 1 in set? Yes, so 2 is not sequence start

Longest sequence: [1,2,3,4] with length 4"
```

#### Phase 6: Intuition Validation
- **Why HashSet works**: Enables O(1) existence checks
- **Why start detection works**: Avoids redundant counting from middle of sequences
- **Why O(N) time**: Each number processed at most twice (once as potential start, once in expansion)
- **Why O(N) space**: HashSet stores all unique numbers

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just sort?"** → Sorting is O(N log N), need O(N) solution
2. **"Should I use Union Find?"** → Can, but HashSet is simpler and more efficient
3. **"What about very large numbers?"** → HashSet handles any integer range
4. **"Can I optimize further?"** → Already optimal O(N) time solution

### Real-World Analogy
**Like finding the longest consecutive streak in attendance records:**
- You have attendance records for various days (unordered dates)
- You want to find the longest streak of consecutive days attended
- Instead of checking every day as a potential start, only start when previous day was absent
- Once you find a streak start, count how many consecutive days follow
- Keep track of the longest streak found

### Human-Readable Pseudocode
```
function longestConsecutive(nums):
    if nums is empty:
        return 0
    
    numSet = set(nums)
    maxLength = 0
    
    for num in numSet:
        if num-1 not in numSet:  // Only start from sequence beginnings
            currentNum = num
            currentLength = 1
            
            while currentNum+1 in numSet:
                currentNum++
                currentLength++
            
            maxLength = max(maxLength, currentLength)
    
    return maxLength
```

### Execution Visualization

### Example Array: [100, 4, 200, 1, 3, 2]
```
HashSet: {1, 2, 3, 4, 100, 200}

Processing:
100: 99 not in set → start sequence [100] → length 1
4:   3 in set → skip
200: 199 not in set → start sequence [200] → length 1  
1:   0 not in set → start sequence [1,2,3,4] → length 4
3:   2 in set → skip
2:   1 in set → skip

Maximum length: 4
```

### Key Visualization Points:
- **HashSet creation**: All numbers stored for O(1) lookup
- **Sequence start detection**: Only process numbers where num-1 doesn't exist
- **Forward expansion**: Count consecutive numbers from each start
- **Maximum tracking**: Keep longest sequence found

### Memory Layout Visualization:
```
HashSet: {1, 2, 3, 4, 100, 200}
             ↓     ↓     ↓
         starts: 1, 100, 200
         sequences: [1,2,3,4], [100], [200]
         max_length: 4
```

### Time Complexity Breakdown:
- **HashSet creation**: O(N) time, O(N) space
- **Sequence detection**: Each number processed at most twice
- **Total time**: O(N) where N is number of elements
- **Space**: O(N) for HashSet storage

### Alternative Approaches:

#### 1. Union Find Approach (O(N α(N)) time, O(N) space)
```go
func longestConsecutiveUnionFind(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    
    // Initialize Union-Find
    parent := make(map[int]int)
    rank := make(map[int]int)
    
    for _, num := range nums {
        parent[num] = num
        rank[num] = 0
    }
    
    var find func(int) int
    find = func(x int) int {
        if parent[x] != x {
            parent[x] = find(parent[x])
        }
        return parent[x]
    }
    
    var union func(int, int)
    union = func(x, y int) {
        rootX := find(x)
        rootY := find(y)
        
        if rootX == rootY {
            return
        }
        
        if rank[rootX] < rank[rootY] {
            parent[rootX] = rootY
        } else if rank[rootX] > rank[rootY] {
            parent[rootY] = rootX
        } else {
            parent[rootY] = rootX
            rank[rootX]++
        }
    }
    
    // Union consecutive numbers
    for _, num := range nums {
        if _, exists := parent[num-1]; exists {
            union(num, num-1)
        }
        if _, exists := parent[num+1]; exists {
            union(num, num+1)
        }
    }
    
    // Count the size of each component
    componentSize := make(map[int]int)
    for _, num := range nums {
        root := find(num)
        componentSize[root]++
    }
    
    maxLength := 0
    for _, size := range componentSize {
        if size > maxLength {
            maxLength = size
        }
    }
    
    return maxLength
}
```
- **Pros**: Demonstrates Union Find pattern, good for dynamic queries
- **Cons**: More complex, higher constant factors

#### 2. Sorting Approach (O(N log N) time, O(1) space)
```go
func longestConsecutiveSorting(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    
    // Remove duplicates and sort
    unique := make(map[int]bool)
    for _, num := range nums {
        unique[num] = true
    }
    
    sorted := make([]int, 0, len(unique))
    for num := range unique {
        sorted = append(sorted, num)
    }
    
    // Sort the array
    sort.Ints(sorted)
    
    maxLength := 1
    currentLength := 1
    
    for i := 1; i < len(sorted); i++ {
        if sorted[i] == sorted[i-1]+1 {
            currentLength++
        } else {
            if currentLength > maxLength {
                maxLength = currentLength
            }
            currentLength = 1
        }
    }
    
    if currentLength > maxLength {
        maxLength = currentLength
    }
    
    return maxLength
}
```
- **Pros**: Simple to understand and implement
- **Cons**: O(N log N) time, not optimal for large inputs

#### 3. Bitset Approach (O(N) time, O(R) space where R is range)
```go
func longestConsecutiveBitset(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    
    // Find min and max to determine range
    minNum, maxNum := nums[0], nums[0]
    for _, num := range nums {
        if num < minNum {
            minNum = num
        }
        if num > maxNum {
            maxNum = num
        }
    }
    
    // Create bitset (simplified - in practice use proper bitset)
    present := make([]bool, maxNum-minNum+1)
    for _, num := range nums {
        present[num-minNum] = true
    }
    
    maxLength := 0
    currentLength := 0
    
    for i := 0; i < len(present); i++ {
        if present[i] {
            currentLength++
            if currentLength > maxLength {
                maxLength = currentLength
            }
        } else {
            currentLength = 0
        }
    }
    
    return maxLength
}
```
- **Pros**: O(N) time, efficient for small number ranges
- **Cons**: O(R) space where R is number range, not suitable for large ranges

### Extensions for Interviews:
- **Longest Consecutive Subarray**: Maintain consecutive in original order
- **K-length Sequences**: Find all sequences of exactly length K
- **Multiple Queries**: Handle multiple longest sequence queries efficiently
- **Circular Array**: Treat array as circular for consecutive sequences
- **Dynamic Updates**: Handle insertions and deletions efficiently
*/
func main() {
	// Test cases
	testCases := [][]int{
		{100, 4, 200, 1, 3, 2},
		{0, 3, 7, 2, 5, 8, 4, 6, 0, 1},
		{},
		{1},
		{1, 2, 0, 1},
		{9, 1, 4, 7, 3, -1, 0, 5, 8, -1, 6},
		{1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
		{10, 9, 8, 7, 6, 5, 4, 3, 2, 1},
		{1, 3, 5, 7, 9},
		{-1, -2, -3, 0, 1, 2, 3},
	}
	
	for i, nums := range testCases {
		result1 := longestConsecutive(nums)
		result2 := longestConsecutiveUnionFind(nums)
		result3 := longestConsecutiveSorting(nums)
		
		fmt.Printf("Test Case %d: %v\n", i+1, nums)
		fmt.Printf("  HashSet: %d, Union-Find: %d, Sorting: %d\n\n", result1, result2, result3)
	}
}

File: long_questions/DSA/Leetcode/java/13_Union_Find/128_Longest_Consecutive_Sequence.java
import java.util.*;

public class LongestConsecutiveSequence {
    
    // 128. Longest Consecutive Sequence
    // Time: O(N), Space: O(N)
    public static int longestConsecutive(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        
        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) {
            numSet.add(num);
        }
        
        int maxLength = 0;
        
        for (int num : numSet) {
            // Only start counting from the beginning of a sequence
            if (!numSet.contains(num - 1)) {
                int currentNum = num;
                int currentLength = 1;
                
                // Count the length of consecutive sequence
                while (numSet.contains(currentNum + 1)) {
                    currentNum++;
                    currentLength++;
                }
                
                maxLength = Math.max(maxLength, currentLength);
            }
        }
        
        return maxLength;
    }

    // Union-Find approach
    public static int longestConsecutiveUnionFind(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        
        // Initialize Union-Find
        Map<Integer, Integer> parent = new HashMap<>();
        Map<Integer, Integer> rank = new HashMap<>();
        
        for (int num : nums) {
            parent.put(num, num);
            rank.put(num, 0);
        }
        
        // Find function with path compression
        class UnionFind {
            int find(int x) {
                if (parent.get(x) != x) {
                    parent.put(x, find(parent.get(x)));
                }
                return parent.get(x);
            }
            
            void union(int x, int y) {
                int rootX = find(x);
                int rootY = find(y);
                
                if (rootX == rootY) {
                    return;
                }
                
                if (rank.get(rootX) < rank.get(rootY)) {
                    parent.put(rootX, rootY);
                } else if (rank.get(rootX) > rank.get(rootY)) {
                    parent.put(rootY, rootX);
                } else {
                    parent.put(rootY, rootX);
                    rank.put(rootX, rank.get(rootX) + 1);
                }
            }
        }
        
        UnionFind uf = new UnionFind();
        
        // Union consecutive numbers
        for (int num : nums) {
            if (parent.containsKey(num - 1)) {
                uf.union(num, num - 1);
            }
            if (parent.containsKey(num + 1)) {
                uf.union(num, num + 1);
            }
        }
        
        // Count the size of each component
        Map<Integer, Integer> componentSize = new HashMap<>();
        for (int num : nums) {
            int root = uf.find(num);
            componentSize.put(root, componentSize.getOrDefault(root, 0) + 1);
        }
        
        int maxLength = 0;
        for (int size : componentSize.values()) {
            maxLength = Math.max(maxLength, size);
        }
        
        return maxLength;
    }

    // Sorting approach (O(N log N))
    public static int longestConsecutiveSorting(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        
        // Remove duplicates and sort
        Set<Integer> unique = new HashSet<>();
        for (int num : nums) {
            unique.add(num);
        }
        
        int[] sorted = new int[unique.size()];
        int index = 0;
        for (int num : unique) {
            sorted[index++] = num;
        }
        
        // Simple bubble sort for demonstration (in practice, use Arrays.sort)
        for (int i = 0; i < sorted.length - 1; i++) {
            for (int j = 0; j < sorted.length - i - 1; j++) {
                if (sorted[j] > sorted[j + 1]) {
                    int temp = sorted[j];
                    sorted[j] = sorted[j + 1];
                    sorted[j + 1] = temp;
                }
            }
        }
        
        int maxLength = 1;
        int currentLength = 1;
        
        for (int i = 1; i < sorted.length; i++) {
            if (sorted[i] == sorted[i - 1] + 1) {
                currentLength++;
            } else {
                maxLength = Math.max(maxLength, currentLength);
                currentLength = 1;
            }
        }
        
        maxLength = Math.max(maxLength, currentLength);
        
        return maxLength;
    }

    public static void main(String[] args) {
        // Test cases
        int[][] testCases = {
            {100, 4, 200, 1, 3, 2},
            {0, 3, 7, 2, 5, 8, 4, 6, 0, 1},
            {},
            {1},
            {1, 2, 0, 1},
            {9, 1, 4, 7, 3, -1, 0, 5, 8, -1, 6},
            {1, 2, 3, 4, 5, 6, 7, 8, 9, 10},
            {10, 9, 8, 7, 6, 5, 4, 3, 2, 1},
            {1, 3, 5, 7, 9},
            {-1, -2, -3, 0, 1, 2, 3}
        };
        
        for (int i = 0; i < testCases.length; i++) {
            int result1 = longestConsecutive(testCases[i]);
            int result2 = longestConsecutiveUnionFind(testCases[i]);
            int result3 = longestConsecutiveSorting(testCases[i]);
            
            System.out.printf("Test Case %d: %s\n", i + 1, Arrays.toString(testCases[i]));
            System.out.printf("  HashSet: %d, Union-Find: %d, Sorting: %d\n\n", 
                result1, result2, result3);
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Union-Find (Disjoint Set Union)
- **Union-Find**: Efficient connectivity data structure
- **Path Compression**: Optimize find operation
- **Union by Rank**: Keep tree shallow
- **Component Analysis**: Track connected component sizes
- **Consecutive Numbers**: Find longest sequence of consecutive integers

## 2. PROBLEM CHARACTERISTICS
- **Connectivity**: Determine if numbers are connected
- **Dynamic Union**: Numbers can be added/connected during processing
- **Component Tracking**: Need to know size of each component
- **Consecutive Constraint**: Numbers must differ by exactly 1

## 3. SIMILAR PROBLEMS
- Number of Provinces
- Graph Valid Tree
- Redundant Connection
- Accounts Merge

## 4. KEY OBSERVATIONS
- Union-Find maintains forest of trees
- Path compression flattens tree structure
- Union by rank prevents tall trees
- Component size tracking enables longest sequence finding
- Time complexity nearly constant for practical purposes

## 5. VARIATIONS & EXTENSIONS
- Union by size instead of rank
- Dynamic connectivity with deletions
- Offline processing of multiple queries
- Different union criteria

## 6. INTERVIEW INSIGHTS
- Clarify: "Are numbers guaranteed to be unique?"
- Edge cases: empty array, single element, all consecutive
- Time complexity: O(α(N)) vs O(N²) naive approach
- Space complexity: O(N) for Union-Find

## 7. COMMON MISTAKES
- Not implementing path compression
- Using union by size instead of rank
- Forgetting to update component sizes
- Off-by-one errors in consecutive checking

## 8. OPTIMIZATION STRATEGIES
- Path compression is crucial for performance
- Union by rank keeps trees balanced
- Lazy union (only when needed)
- Component size tracking for longest sequence

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like managing social networks:**
- Each person is a node in a social network
- Union-Find tracks who's connected to whom
- When two people become friends, their networks merge
- Path compression finds the "ultimate friend" quickly
- Component size tells you how big each friend group is

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of integers
2. **Goal**: Find length of longest consecutive sequence
3. **Output**: Maximum length of consecutive numbers

#### Phase 2: Key Insight Recognition
- **"What makes numbers consecutive?"** → Numbers that differ by exactly 1
- **"How to track connectivity?"** → Union-Find data structure
- **"What's the key insight?"** → Consecutive numbers must be in same component
- **"How to find longest?"** → Track maximum component size

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use Union-Find:
1. For each number, create its own component initially
2. For consecutive numbers, union their components
3. Track size of each component
4. The maximum component size is the answer"
```

#### Phase 4: Edge Case Handling
- **Empty array**: Return 0
- **Single element**: Return 1
- **All consecutive**: Return array length
- **No consecutive pairs**: Return 1

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Numbers: [100, 4, 200, 1, 3, 2]

Human thinking:
"Initialize Union-Find with each number in its own set:
- 100: {100}, 4: {4}, 200: {200}, 1: {1}, 3: {3}, 2: {2}

Check consecutive pairs:
- 100 and 99? 99 not in array
- 100 and 101? 101 not in array
- 4 and 3? Both exist, union(4,3)
- 200 and 199? 199 not in array
- 200 and 201? 201 not in array
- 1 and 0? 0 not in array
- 1 and 2? Both exist, union(1,2)
- 3 and 2? Both exist, union(3,2)
- 3 and 4? Both exist, union(3,4)

Components after unions:
- {100}, {4,3}, {200}, {1,2}, {3,2}

Component sizes:
- {100}: size 1
- {4,3}: size 2
- {200}: size 1
- {1,2}: size 2
- {3,2}: size 2

Maximum size: 2 ✓"
```

#### Phase 6: Intuition Validation
- **Why it works**: Union-Find correctly tracks connectivity
- **Why it's efficient**: Path compression makes operations nearly O(1)
- **Why it's correct**: Consecutive numbers must be connected

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just sort and scan?"** → O(N log N) but Union-Find is more elegant
2. **"What about path compression?"** → Essential for performance
3. **"How to track component sizes?"** → Update during union operations
4. **"What about duplicate numbers?"** → Need to handle properly

### Real-World Analogy
**Like managing company departments:**
- Each employee starts in their own department (individual component)
- When employees collaborate, their departments merge
- Union-Find tracks which department each employee belongs to
- Path compression helps find the "ultimate department" quickly
- Component size shows how many employees are in each department
- Largest department size is like the biggest team collaboration

### Human-Readable Pseudocode
```
function longestConsecutive(nums):
    if nums is empty:
        return 0
    
    uf = UnionFind()
    
    // Initialize each number as its own component
    for num in nums:
        uf.makeSet(num)
    
    // Union consecutive numbers
    for num in nums:
        if uf.contains(num - 1):
            uf.union(num, num - 1)
        if uf.contains(num + 1):
            uf.union(num, num + 1)
    
    // Find maximum component size
    maxSize = 0
    for num in nums:
        root = uf.find(num)
        size = uf.getComponentSize(root)
        maxSize = max(maxSize, size)
    
    return maxSize
```

### Execution Visualization

### Example: nums = [100, 4, 200, 1, 3, 2]
```
Initial: Each number in its own component
Components: {100}, {4}, {200}, {1}, {3}, {2}

Union consecutive pairs:
- 4 and 3: union(4,3) → {4,3}
- 1 and 2: union(1,2) → {1,2}
- 3 and 2: union(3,2) → {1,2,3,2}

Final components:
- {100}: size 1
- {4,3,200}: size 3
- {1,2,3,2}: size 4

Maximum size: 4 ✓

Longest consecutive sequence: [1,2,3,2] or [3,2,4,3] (length 4)"
```

### Key Visualization Points:
- **Union-Find** maintains dynamic connectivity
- **Path compression** optimizes find operations
- **Component tracking** enables size queries
- **Consecutive checking** unites appropriate numbers
- **Maximum size** gives longest sequence length

### Memory Layout Visualization:
```
Union-Find Forest Evolution:
Initial:  {100} {4} {200} {1} {3} {2}
         |     |     |      |     |     |

After unions:
 {100} {4,3,200} {1,2,3,2}
         |     \     |      /     |
         |     |     |      |

Component Roots:
100 → rootA (size 1)
4,3,200 → rootB (size 3)
1,2,3,2 → rootC (size 4)

Path Compression:
find(3) → rootC (direct)
find(1) → rootC (direct)
```

### Time Complexity Breakdown:
- **Make Set**: O(N) operations
- **Union Operations**: O(α(N)) where α is inverse Ackermann function
- **Find Operations**: O(α(N)) with path compression
- **Component Tracking**: O(N) additional space
- **Total**: O(N) time for practical purposes, O(N) space
- **Optimal**: Nearly constant time per operation
*/
